go/vt/vtgate/planbuilder/union.go

```diff
@@ -0,0 +1,112 @@
+package planbuilder
+
+import (
+	"errors"
+
+	"github.com/youtube/vitess/go/vt/sqlparser"
+	"github.com/youtube/vitess/go/vt/vtgate/engine"
+	"github.com/youtube/vitess/go/vt/vtgate/vindexes"
+)
+
+func buildUnionPlan(union *sqlparser.Union, vschema VSchema) (primitive engine.Primitive, err error) {
+	bindvars := sqlparser.GetBindvars(union)
+	bldr, err := processUnion(union, vschema, nil)
+	if err != nil {
+		return nil, err
+	}
+	jt := newJointab(bindvars)
+	err = bldr.Wireup(bldr, jt)
+	if err != nil {
+		return nil, err
+	}
+	return bldr.Primitive(), nil
+}
+
+func processUnion(union *sqlparser.Union, vschema VSchema, outer builder) (builder, error) {
+	lbldr, err := processPart(union.Left, vschema, outer)
+	if err != nil {
+		return nil, err
+	}
+	rbldr, err := processPart(union.Right, vschema, outer)
+	if err != nil {
+		return nil, err
+	}
+	bldr, err := unionRouteMerge(union, lbldr, rbldr, vschema)
+	if err == nil {
+		return bldr, nil
+	}
+	if union.Type != sqlparser.UnionAllStr {
+		return nil, errors.New("unsupported: multi-shard union without ALL")
+	}
+	//TODO(acharis): need to check that lhs and rhs have same number of columns
+	return newUnionBuilder(lbldr, rbldr)
+}
+
+func processPart(part sqlparser.SelectStatement, vschema VSchema, outer builder) (builder, error) {
+	var err error
+	var bldr builder
+	switch part := part.(type) {
+	case *sqlparser.Union:
+		bldr, err = processUnion(part, vschema, outer)
+	case *sqlparser.Select:
+		bldr, err = processSelect(part, vschema, outer)
+	case *sqlparser.ParenSelect:
+		bldr, err = processPart(part.Select, vschema, outer)
```

Comment:
You may need to re-parenthesize the statement. Otherwise, the generated code will lose the parenthesis. So, `select...  union (select... limit..)` would become `select... union select... limit...`, which means something else. Maybe a good idea to add a test case.